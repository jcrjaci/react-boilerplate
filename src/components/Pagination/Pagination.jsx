import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import './Pagination.scss';

class Pagination extends Component {
  static propTypes = {
    currentPage: PropTypes.number.isRequired,
    totalPages: PropTypes.number.isRequired,
  };

  pages = 3;

  getPageLink = (current, page, text = page) => <Link className={`${current === page ? 'active' : ''}`} key={page} to={`/coins/${page}`}>{text}</Link>;

  getBeforePages = (current) => {
    // Number of before pages is 3 unless the current page is lower than 3
    const beforePages = current > this.pages ? this.pages : current - 1;
    // create an array with the number of before pages
    const pages = Array.from({ length: beforePages }, (v, i) => i).sort((a, b) => a < b);

    return pages.map(page => this.getPageLink(current, current - page - 1 ));
  }

  getAfterPages = (current, total) => {
    // Number of after pages is 3 unless the current page more  3 is higher than the total pages
    const afterPages = total >= current + this.pages ? this.pages : total - current;
    // create an array with the number of after pages
    const pages = Array.from({ length: afterPages }, (v, i) => i);

    return pages.map(page => this.getPageLink(current, current + page + 1));
  }

  getPagesLinks = (current, total) => {
    const beforePages = current > 1 ? this.getBeforePages(current) : [];
    const currentPageLink = this.getPageLink(current, current);
    const afterPages = current <= total ? this.getAfterPages(current, total) : [];

    return [...beforePages, currentPageLink, ...afterPages];
  }

  getPreviousPageLink = currentPage => (currentPage > 1 ? this.getPageLink(currentPage, currentPage - 1, '\u276E') : null);

  getNextPageLink = (currentPage, totalPages) => (currentPage !== totalPages ? this.getPageLink(currentPage, currentPage + 1, '\u276F') : null);

  // https://codepen.io/andrewichert/pen/bZXBbO
  render() {
    const { currentPage, totalPages } = this.props;
    const renderPrevious = currentPage > 1;
    const renderLast = currentPage !== totalPages;

    return (
      <div className="pagination">
        {this.getPreviousPageLink(currentPage)}
        {renderPrevious && this.getPageLink(currentPage, 1, 1)}
        <span>...</span>
        {this.getPagesLinks(currentPage, totalPages)}
        <span>...</span>
        {renderLast && this.getPageLink(currentPage, totalPages, totalPages)}
        {this.getNextPageLink(currentPage)}
      </div>
    );
  }
}

export default Pagination;
