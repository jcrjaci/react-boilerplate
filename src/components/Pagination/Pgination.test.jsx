import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { configure, shallow } from 'enzyme';
import { Link } from 'react-router-dom';
import Pagination from './Pagination';

configure({ adapter: new Adapter() });

describe('<Pagination />', () => {
  const wrapper = shallow(<Pagination currentPage={10} totalPages={100} />);

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should only render "1"', () => {
    wrapper.setProps({ currentPage: 1, totalPages: 1 });
    const link = wrapper.find(Link);

    expect(link).toHaveLength(1);

    expect(link.get(0).props.children).toBe(1);
    expect(link.get(0).props.className).toBe('active');
    expect(link.get(0).props.to).toBe('/coins/1');
  });

  it('should only render "1 2 >" with 1 as active', () => {
    wrapper.setProps({ currentPage: 1, totalPages: 2 });
    const link = wrapper.find(Link);

    expect(link).toHaveLength(3);

    expect(link.get(0).props.children).toBe(1);
    expect(link.get(0).props.className).toBe('active');
    expect(link.get(0).props.to).toBe('/coins/1');

    expect(link.get(1).props.children).toBe(2);
    expect(link.get(1).props.to).toBe('/coins/2');

    expect(link.get(2).props.children).toBe('\u276F');
    expect(link.get(2).props.className).toBe('next');
    expect(link.get(2).props.to).toBe('/coins/2');
  });

  it('should only render "< 1 2" with 2 as active', () => {
    wrapper.setProps({ currentPage: 2, totalPages: 2 });
    const link = wrapper.find(Link);

    expect(link).toHaveLength(3);
    expect(link.get(0).props.to).toBe('/coins/1');
    expect(link.get(0).props.children).toBe('\u276E');
    expect(link.get(0).props.className).toBe('prev');

    expect(link.get(1).props.children).toBe(1);
    expect(link.get(1).props.to).toBe('/coins/1');

    expect(link.get(2).props.children).toBe(2);
    expect(link.get(2).props.to).toBe('/coins/2');
    expect(link.get(2).props.className).toBe('active');
  });

  it('should only render "< 1 2 3 >" with 2 as active', () => {
    wrapper.setProps({ currentPage: 2, totalPages: 3 });
    const link = wrapper.find(Link);

    expect(link).toHaveLength(5);

    expect(link.get(0).props.to).toBe('/coins/1');
    expect(link.get(0).props.children).toBe('\u276E');
    expect(link.get(0).props.className).toBe('prev');

    expect(link.get(1).props.children).toBe(1);
    expect(link.get(1).props.to).toBe('/coins/1');

    expect(link.get(2).props.children).toBe(2);
    expect(link.get(2).props.to).toBe('/coins/2');
    expect(link.get(2).props.className).toBe('active');

    expect(link.get(3).props.children).toBe(3);
    expect(link.get(3).props.to).toBe('/coins/3');

    expect(link.get(4).props.children).toBe('\u276F');
    expect(link.get(4).props.className).toBe('next');
    expect(link.get(4).props.to).toBe('/coins/3');
  });

  it('should only render "1 2 3 >" with 1 as active', () => {
    wrapper.setProps({ currentPage: 1, totalPages: 3 });
    const link = wrapper.find(Link);

    expect(link).toHaveLength(4);

    expect(link.get(0).props.children).toBe(1);
    expect(link.get(0).props.to).toBe('/coins/1');
    expect(link.get(0).props.className).toBe('active');

    expect(link.get(1).props.children).toBe(2);
    expect(link.get(1).props.to).toBe('/coins/2');

    expect(link.get(2).props.children).toBe(3);
    expect(link.get(2).props.to).toBe('/coins/3');

    expect(link.get(3).props.children).toBe('\u276F');
    expect(link.get(3).props.className).toBe('next');
    expect(link.get(3).props.to).toBe('/coins/2');
  });

  it('should only render "< 1 2 3" with 3 as active', () => {
    wrapper.setProps({ currentPage: 3, totalPages: 3 });
    const link = wrapper.find(Link);

    expect(link).toHaveLength(4);

    expect(link.get(0).props.to).toBe('/coins/2');
    expect(link.get(0).props.children).toBe('\u276E');
    expect(link.get(0).props.className).toBe('prev');

    expect(link.get(1).props.children).toBe(1);
    expect(link.get(1).props.to).toBe('/coins/1');

    expect(link.get(2).props.children).toBe(2);
    expect(link.get(2).props.to).toBe('/coins/2');

    expect(link.get(3).props.children).toBe(3);
    expect(link.get(3).props.to).toBe('/coins/3');
    expect(link.get(3).props.className).toBe('active');
  });

  it('should only render "< 1 2 3 4 >" with 3 as active', () => {
    wrapper.setProps({ currentPage: 3, totalPages: 4 });
    const link = wrapper.find(Link);

    expect(link).toHaveLength(6);

    expect(link.get(0).props.to).toBe('/coins/2');
    expect(link.get(0).props.children).toBe('\u276E');
    expect(link.get(0).props.className).toBe('prev');

    expect(link.get(1).props.children).toBe(1);
    expect(link.get(1).props.to).toBe('/coins/1');

    expect(link.get(2).props.children).toBe(2);
    expect(link.get(2).props.to).toBe('/coins/2');

    expect(link.get(3).props.children).toBe(3);
    expect(link.get(3).props.to).toBe('/coins/3');
    expect(link.get(3).props.className).toBe('active');

    expect(link.get(4).props.children).toBe(4);
    expect(link.get(4).props.to).toBe('/coins/4');

    expect(link.get(5).props.children).toBe('\u276F');
    expect(link.get(5).props.className).toBe('next');
    expect(link.get(5).props.to).toBe('/coins/4');
  });

  it('should only render "< 1 2 3 4 >" with 2 as active', () => {
    wrapper.setProps({ currentPage: 2, totalPages: 4 });
    const link = wrapper.find(Link);

    expect(link).toHaveLength(6);

    expect(link.get(0).props.to).toBe('/coins/1');
    expect(link.get(0).props.children).toBe('\u276E');
    expect(link.get(0).props.className).toBe('prev');

    expect(link.get(1).props.children).toBe(1);
    expect(link.get(1).props.to).toBe('/coins/1');

    expect(link.get(2).props.children).toBe(2);
    expect(link.get(2).props.to).toBe('/coins/2');
    expect(link.get(2).props.className).toBe('active');

    expect(link.get(3).props.children).toBe(3);
    expect(link.get(3).props.to).toBe('/coins/3');

    expect(link.get(4).props.children).toBe(4);
    expect(link.get(4).props.to).toBe('/coins/4');

    expect(link.get(5).props.children).toBe('\u276F');
    expect(link.get(5).props.className).toBe('next');
    expect(link.get(5).props.to).toBe('/coins/3');
  });

  it('should only render "< 1 ...  3 4 5 >" with 4 as active', () => {
    wrapper.setProps({ currentPage: 4, totalPages: 5 });
    const link = wrapper.find(Link);
    const span = wrapper.find('span');

    expect(link).toHaveLength(6);
    expect(span).toHaveLength(1);

    expect(link.get(0).props.to).toBe('/coins/3');
    expect(link.get(0).props.children).toBe('\u276E');
    expect(link.get(0).props.className).toBe('prev');

    expect(link.get(1).props.children).toBe(1);
    expect(link.get(1).props.to).toBe('/coins/1');

    expect(link.get(2).props.children).toBe(3);
    expect(link.get(2).props.to).toBe('/coins/3');

    expect(link.get(3).props.children).toBe(4);
    expect(link.get(3).props.to).toBe('/coins/4');
    expect(link.get(3).props.className).toBe('active');

    expect(link.get(4).props.children).toBe(5);
    expect(link.get(4).props.to).toBe('/coins/5');

    expect(link.get(5).props.children).toBe('\u276F');
    expect(link.get(5).props.className).toBe('next');
    expect(link.get(5).props.to).toBe('/coins/5');
  });

  it('should only render "< 1 2 3 4 ... 6 >" with 3 as active', () => {
    wrapper.setProps({ currentPage: 3, totalPages: 6 });
    const link = wrapper.find(Link);
    const span = wrapper.find('span');

    expect(link).toHaveLength(7);
    expect(span).toHaveLength(1);

    expect(link.get(0).props.to).toBe('/coins/2');
    expect(link.get(0).props.children).toBe('\u276E');
    expect(link.get(0).props.className).toBe('prev');

    expect(link.get(1).props.children).toBe(1);
    expect(link.get(1).props.to).toBe('/coins/1');

    expect(link.get(2).props.children).toBe(2);
    expect(link.get(2).props.to).toBe('/coins/2');

    expect(link.get(3).props.children).toBe(3);
    expect(link.get(3).props.to).toBe('/coins/3');
    expect(link.get(3).props.className).toBe('active');

    expect(link.get(4).props.children).toBe(4);
    expect(link.get(4).props.to).toBe('/coins/4');

    expect(link.get(5).props.children).toBe(6);
    expect(link.get(5).props.to).toBe('/coins/6');

    expect(link.get(6).props.children).toBe('\u276F');
    expect(link.get(6).props.className).toBe('next');
    expect(link.get(6).props.to).toBe('/coins/4');
  });

  it('should only render "< 1 ... 3 4 5 ... 7 >" with 4 as active', () => {
    wrapper.setProps({ currentPage: 4, totalPages: 7 });
    const link = wrapper.find(Link);
    const span = wrapper.find('span');

    expect(link).toHaveLength(7);
    expect(span).toHaveLength(2);

    expect(link.get(0).props.to).toBe('/coins/3');
    expect(link.get(0).props.children).toBe('\u276E');
    expect(link.get(0).props.className).toBe('prev');

    expect(link.get(1).props.children).toBe(1);
    expect(link.get(1).props.to).toBe('/coins/1');

    expect(link.get(2).props.children).toBe(3);
    expect(link.get(2).props.to).toBe('/coins/3');

    expect(link.get(3).props.children).toBe(4);
    expect(link.get(3).props.to).toBe('/coins/4');
    expect(link.get(3).props.className).toBe('active');

    expect(link.get(4).props.children).toBe(5);
    expect(link.get(4).props.to).toBe('/coins/5');

    expect(link.get(5).props.children).toBe(7);
    expect(link.get(5).props.to).toBe('/coins/7');

    expect(link.get(6).props.children).toBe('\u276F');
    expect(link.get(6).props.className).toBe('next');
    expect(link.get(6).props.to).toBe('/coins/5');
  });
});
