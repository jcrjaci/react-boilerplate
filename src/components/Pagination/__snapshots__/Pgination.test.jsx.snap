// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should render correctly 1`] = `
<div
  className="pagination"
>
  <Link
    className="prev"
    key="prev"
    replace={false}
    to="/coins/9"
  >
    ❮
  </Link>
  <Link
    key="1"
    replace={false}
    to="/coins/1"
  >
    1
  </Link>
  <span>
    ...
  </span>
  <Link
    key="9"
    replace={false}
    to="/coins/9"
  >
    9
  </Link>
  <Link
    className="active"
    key="10"
    replace={false}
    to="/coins/10"
  >
    10
  </Link>
  <Link
    key="11"
    replace={false}
    to="/coins/11"
  >
    11
  </Link>
  <span>
    ...
  </span>
  <Link
    key="100"
    replace={false}
    to="/coins/100"
  >
    100
  </Link>
  <Link
    className="next"
    key="next"
    replace={false}
    to="/coins/11"
  >
    ❯
  </Link>
</div>
`;
